### PR TITLE
Genquery1 enhancements and touch operation for /collections

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: test/test_irods_http_api.py [hunk 1]
```diff
@@ -1107,6 +1107,111 @@ def test_support_for_genquery1(self):
...
+        # Create a data object with mixed case letters in the name.
+        data_object = os.path.join('/', self.zone_name, 'home', self.rodsuser_username, 'test_GenQuery1_NO_DISTINCT.TxT')
+        r = requests.post(f'{self.url_base}/data-objects', headers=headers, data={'op': 'touch', 'lpath': data_object})
+        #print(r.content) # Debug
+        self.assertEqual(r.status_code, 200)
+        result = r.json()
+        self.assertEqual(result['irods_response']['status_code'], 0)
+
+        coll_name = os.path.dirname(data_object)
+        data_name = os.path.basename(data_object)
```
Comment (trel) [inline test/test_irods_http_api.py:1263]: do you need a `lower()` here or something to make sure you're searching for something different than is in the catalog?

File: test/test_irods_http_api.py [hunk 2]
```diff
@@ -439,6 +439,82 @@ def test_modifying_permissions_atomically(self):
...
+
+            # Show an error is returned when the touch operation for a collection is given
+            # a path pointing to a data object.
+            r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
+                'op': 'touch',
+                'lpath': data_object
+            })
+            #print(r.content) # Debug
+            self.assertEqual(r.status_code, 400)
+            self.assertEqual(r.json()['irods_response']['status_code'], -170000)
```
Comment (trel) [inline test/test_irods_http_api.py:507]: add comment that this is `NOT_A_COLLECTION`

File: test/test_irods_http_api.py [hunk 3]
```diff
@@ -439,6 +439,82 @@ def test_modifying_permissions_atomically(self):
...
+                'op': 'touch',
+                'lpath': data_object
+            })
+            #print(r.content) # Debug
+            self.assertEqual(r.status_code, 200)
+            self.assertEqual(r.json()['irods_response']['status_code'], 0)
+
+            # Show an error is returned when the touch operation for a collection is given
+            # a path pointing to a data object.
+            r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
```
Comment (trel) [inline test/test_irods_http_api.py:501]: is this correct?    

`self.url_endpoint` ?

Comment (korydraughn) [reply test/test_irods_http_api.py:501]: Yes. `self.url_endpoint` is defined for the primary endpoint we're testing under. In this case, the `/collections` endpoint.

Comment (trel) [reply test/test_irods_http_api.py:501]: okay.  thanks.

File: test/test_irods_http_api.py (hunk 1 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:1328]: Done.

File: test/test_irods_http_api.py (hunk 2 above)
Comment (korydraughn) [reply test/test_irods_http_api.py]: Done.

**Comment by korydraughn** [comment]
Added a new test for the touch op of /data-objects.

All tests pass.

File: src/endpoints/query/impl.cpp [hunk 4]
```diff
@@ -261,10 +261,20 @@ namespace
 								options |= UPPER_CASE_WHERE;
 								boost::algorithm::to_upper(query_iter->second);
 							}
+							else if (iter->second != "1") {
+								log::error("{}: Invalid value for [case-sensitive] parameter. Expected a 1 or 0.", fn);
```
Comment (trel) [inline src/endpoints/query/impl.cpp:265]: want to log the invalid value used?  not sure on this...

File: src/endpoints/query/impl.cpp [hunk 5]
```diff
@@ -261,10 +261,20 @@ namespace
...
 						}
 
-						if (const auto iter = args.find("distinct"); iter != std::end(args) && iter->second == "0") {
-							options |= NO_DISTINCT;
+						if (const auto iter = args.find("distinct"); iter != std::end(args)) {
+							if (iter->second == "0") {
+								options |= NO_DISTINCT;
+							}
+							else if (iter->second != "1") {
+								log::error("{}: Invalid value for [distinct] parameter. Expected a 1 or 0.", fn);
```
Comment (trel) [inline src/endpoints/query/impl.cpp:275]: want to log the invalid value used? not sure on this...

File: src/endpoints/query/impl.cpp (hunk 4 above)
Comment (korydraughn) [reply src/endpoints/query/impl.cpp:265]: I'm going to do a pass over all log messages in the future.

I'll create an issue for this.

Comment (korydraughn) [reply src/endpoints/query/impl.cpp:265]: Created #157 for this.

File: test/test_irods_http_api.py [hunk 6]
```diff
@@ -1107,6 +1107,131 @@ def test_support_for_genquery1(self):
...
+        finally:
+            # Remove data object.
+            r = requests.post(f'{self.url_base}/data-objects', headers=headers, data={
+                'op': 'remove',
+                'lpath': data_object,
+                'no-trash': 1
+            })
+            #print(r.content) # Debug
+
+    @unittest.skip('Test needs to be implemented.')
```
Comment (alanking) [inline test/test_irods_http_api.py:1371]: Got an issue for this?

File: test/test_irods_http_api.py [hunk 7]
```diff
@@ -1107,6 +1107,131 @@ def test_support_for_genquery1(self):
...
+            # Remove resource.
+            r = requests.post(f'{self.url_base}/resources', headers=rodsadmin_headers, data={
+                'op': 'remove',
+                'name': resc_name
+            })
+            #print(r.content) # Debug
+
+    def test_genquery1_case_sensitivity_option(self):
+        headers = {'Authorization': 'Bearer ' + self.rodsuser_bearer_token}
+        data_object = os.path.join('/', self.zone_name, 'home', self.rodsuser_username, 'test_GenQuery1_NO_DISTINCT.TxT')
```
Comment (alanking) [inline test/test_irods_http_api.py:1319]:
```suggestion
        data_object = os.path.join('/', self.zone_name, 'home', self.rodsuser_username, 'test_GenQuery1_CASE_SENSITIVE.TxT')
```

File: test/test_irods_http_api.py [hunk 8]
```diff
@@ -439,6 +439,82 @@ def test_modifying_permissions_atomically(self):
...
+            r = requests.post(f'{self.url_base}/data-objects', headers=rodsuser_headers, data={
+                'op': 'touch',
+                'lpath': data_object
+            })
+            #print(r.content) # Debug
+            self.assertEqual(r.status_code, 200)
+            self.assertEqual(r.json()['irods_response']['status_code'], 0)
+
+            # Show an error is returned when the touch operation for a collection is given
+            # a path pointing to a data object.
```
Comment (alanking) [inline test/test_irods_http_api.py:500]: Should we add a test for touching a path which does not exist?

**Review by alanking — COMMENTED** [review]
Do we need a bit in the README for the new touch operation for collections?

File: test/test_irods_http_api.py (hunk 6 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:1371]: Created #158 for this.

File: test/test_irods_http_api.py (hunk 8 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:500]: That makes sense.

Will add one for /collections.

**Comment by korydraughn** [comment]
> Do we need a bit in the README for the new touch operation for collections?

Yep. Forgot to update the API documentation.

File: test/test_irods_http_api.py (hunk 7 above)
Comment (korydraughn) [reply test/test_irods_http_api.py]: Done.

File: test/test_irods_http_api.py (hunk 8 above)
Comment (korydraughn) [reply test/test_irods_http_api.py:500]: Done.

File: API.md [hunk 9]
```diff
@@ -401,6 +401,36 @@ If an HTTP status code of 200 is returned, the body of the response will contain
...
+```js
+{
+    "irods_response": {
+        "status_code": 0
+        "status_message": "string" // Optional
+    }
+}
+```
+
+If there was an error, expect either an HTTP status code in the 4XX or 5XX range.
```
Comment (trel) [inline API.md:432]: update
"either an HTTP status code in"
to
"an HTTP status code in either"

Comment (korydraughn) [reply API.md]: Done.

**Comment by korydraughn** [comment]
All code review comments have been resolved.
All tests pass.

I think this is ready for squashing and merging.

File: test/test_irods_http_api.py [hunk 10]
```diff
@@ -439,6 +439,104 @@ def test_modifying_permissions_atomically(self):
...
+        # does not exist. This is intended.
+        r = requests.post(self.url_endpoint, headers=rodsuser_headers, data={
+            'op': 'touch',
+            'lpath': f'/{self.zone_name}/home/{self.rodsuser_username}/does_not_exist'
+        })
+        #print(r.content) # Debug
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.json()['irods_response']['status_code'], 0)
+
+        # Show the touch operation did not create a new collection or data object (very bad).
```
Comment (trel) [inline test/test_irods_http_api.py:531]: '(very bad)'
to
'(would be very bad)'

**Comment by trel** [comment]
agreed - one last tweak above, and then squash to ... three commits?

File: test/test_irods_http_api.py (hunk 10 above)
Comment (korydraughn) [reply test/test_irods_http_api.py]: Done.

**Comment by korydraughn** [comment]
Squashed to four commits to avoid hiding when the touch op test for /data-objects was added.

**Comment by korydraughn** [comment]
#'d

**Pull request merged**